### PR TITLE
Fix search and status in Exchange

### PR DESCRIPTION
**Pull request by armiantos**
# Summary
- Search previously includes all books that have that title, now client filters it to only include books that are requestable (not borrowed or accepted, and not yours)
  - Trade off of not storing status and owner in algolia while minimizing search requests in server
  - Assumes not many of the book requests will be discarded
- ObjectID in Algolia now matches book ID in Firestore to eliminate redundancy

Solves half of #122

### CODE REVIEW

File: app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java [hunk 1]
```diff
@@ -50,14 +50,15 @@
 
     // Algolia fields
     private static final String ALGOLIA_INDEX_NAME = "books";
-    private static final String ALGOLIA_ID_FIELD = "id";
+    private static final String ALGOLIA_ID_FIELD = "objectID";
```
Comment (calebschoepp) [inline app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java:53]: Why is this changing, just curious.

File: app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java [hunk 2]
```diff
@@ -122,9 +127,24 @@ public void attachListener() {
...
+     *
+     * @param request           request object to be saved in the database
+     * @param onSuccessListener code to call on success
+     * @param onFailureListener code to call on failure
+     */
+    public void createRequest(Request request, OnSuccessListener<DocumentReference> onSuccessListener, OnFailureListener onFailureListener) {
+        mDb.collection(REQUEST_COLLECTION).add(request)
+                .addOnSuccessListener(onSuccessListener)
+                .addOnFailureListener(onFailureListener);
+    }
```
Comment (calebschoepp) [inline app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java:140]: Why is this being added in this PR? It seems unrelated.

File: app/src/main/java/com/example/unlibrary/library/LibraryRepository.java [hunk 3]
```diff
@@ -170,6 +170,13 @@ public void deleteBook(Book book, OnSuccessListener<? super Void> onSuccessListe
                 .delete()
                 .addOnSuccessListener(onSuccessListener)
                 .addOnFailureListener(onFailureListener);
+
+        mAlgoliaClient.getIndex(ALGOLIA_INDEX_NAME)
+                .deleteObjectAsync(book.getId(), (jsonObject, e) -> {
+                    if (e != null) {
+                        Log.e(TAG, "Unable to delete book " + book.getId(), e);
+                    }
+                });
```
Comment (calebschoepp) [inline app/src/main/java/com/example/unlibrary/library/LibraryRepository.java:179]: If I understand correctly, this is deleting a book from the Algolia index when it is deleted by the user, right?

**Review by calebschoepp — COMMENTED** [review]
Good job man. I just have a couple of questions.

File: app/src/main/java/com/example/unlibrary/library/LibraryRepository.java (hunk 3 above)
Comment (armiantos) [reply app/src/main/java/com/example/unlibrary/library/LibraryRepository.java:179]: Yep!

File: app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java (hunk 2 above)
Comment (armiantos) [reply app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java:140]: Ah wait I think this was a rebase let me remove it

File: app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java (hunk 1 above)
Comment (armiantos) [reply app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java:53]: Previously algolia we had 4 fields
```json
{
    "objectID": "automatically assigned by Algolia",
    "id": "Firestore reference",
    "title": "Book title",
    "author": "Book author"
}
```
In order to delete efficiently, we need the Algolia objectID. This is to save search requests on Algolia (we don't need to query for object with given Firestore ID. Searching for matching Firestore ID is treated the same way as full text search otherwise)

**Review by armiantos — COMMENTED** [review]
Replied to some comments, will update the misincluded method

File: app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java (hunk 2 above)
Comment (armiantos) [reply app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java]: Fixed!

**Comment by calebschoepp** [comment]
> In order to delete efficiently, we need the Algolia objectID. This is to save search requests on Algolia (we don't need to query for object with given Firestore ID. Searching for matching Firestore ID is treated the same way as full text search otherwise)

Thanks for explaining, cool to learn.

**Comment by armiantos** [comment]
@calebschoepp lmk when this is good to merge

**Review by calebschoepp — APPROVED** [review]
LGTM! :ship:
